src/wallet/wallet.cpp

```diff
@@ -1206,50 +1227,100 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
         pblock = &block;
     }
 
-    for (const CTransaction& tx : pblock->vtx) {
-        auto hash = tx.GetHash();
-        bool txIsOurs = mapWallet.count(hash);
+    // 1) Loop over the block txs and gather the note commitments ordered.
+    // If the tx is from this wallet, witness it and append the next block note commitments on top.
+    std::vector<uint256> noteCommitmentsSapling;
+    std::vector<uint256> noteCommitmentsSprout;
+    std::vector<std::pair<CWalletTx*, SaplingNoteData*>> inBlockNotesSap;
+    std::vector<std::pair<CWalletTx*, SproutNoteData*>> inBlockNotesSprout;
+    for (const auto& tx : pblock->vtx) {
+        if (!tx.IsShieldTx()) continue;
```

Comment:
```suggestion
        if (!tx.HasShieldedComponents()) continue;
```
This PR will also need other nontrivial changes to accommodate Orchard.